README.md
Node.js Command Line App to get Weather from Forecast.io
=================

Based on the Node.js Command Line App for retrieving account badge and points info from [Treehouse](http://teamtreehouse.com) in JSON format from [@Chalkers](http://twitter.com/chalkers/). Takes ```process.argv.slice(2)``` from command line for zip code for current weather forecast and returns from [API at Forecast.io](https://developer.forecast.io/).

### Version

v0.1.3

### To run —

You will need [Node.JS ≥ v0.10.36](http://nodejs.org "The Node.JS Main Web Site.") and the [async](https://github.com/caolan/async) module from [npm](http://npmjs.org "The NPM Main Web Site."). Once all that and this repo is loaded, Just type from command line: ```$ node forecast.js [U.S. ZIPCODE(S)]```, as in: ```$ node forecast.js 10010```; or several zip codes, as thus: ```$ node forecast.js 10010 20050 94101 98109 97210```. It all works! Like magic!

### Current Info/Still Needed

* Separated into 4 main files —
  * forecast.js — main command line app file run from node
  * zip-convert.js — takes u.s. zip from command line argv and converts it to longitude and latitude for [forecast.io](http://forecast.io/) processing.
  * zip-forecast.js — takes converted from zip code longitude/latitude from zip-convert.js and uses forecast.io API and KEY to return weather forecast.
  * messages.js — contains the messages used in the other *.js files to printLocationInfo(), printForecastInfo() & printError().
* All is now completely working! Just type from command line: ```$ node forecast.js [U.S. ZIPCODE(S)]```, as in: ```$ node forecast.js 10010```; or several zip codes, as thus: ```$ node forecast.js 10010 20050 94101 98109 97210```. It all works! Like magic!
* New feature — if location has special weather alert, messages.js printForecastMessage() now prints that out.
* Now fully converting Atmospheric Pressure in hPA to more common U.S. Barometric Pressure inHG (inches Mercury) (from the hPA to inHG conversion at the [conversion units site](http://www.convertunits.com/from/hpa/to/inhg).
* Humidity now showing as percentage.
* Unicode string chars now printing out ok in terminal CLI, just need to make sure character encoding in terminal is set always to UTF.
* forecast.js now has optional forEach() code to call other module callback functions *(commented out; it's faster, but does not work as well as async.eachSeries())*.
* Added in a number of appropriate line breaks in console.log() print outs and and corrected error messages for servers so they are now consistent for both forecast.io and geonames.org.
* As of evening 25-Feb.-2015, added auto-generated CHANGELOG.md generated by npm module github-changes. Further changes/additions found in there.
* Start creating a front-end, perhaps in Angular.JS with some Twitter-Bootstrap — get coding!
* Future version ideas — maybe other country postal codes also?

### Required Software

* [Node.JS ≥ v0.10.36](http://nodejs.org "The Node.JS Main Web Site.")
* [async](https://github.com/caolan/async) module from npm
* [Forecast.io API and API KEY](https://developer.forecast.io/)
* [Geonames API Account](http://www.geonames.org/) that is [Enabled for your Log-in User Name](http://www.geonames.org/manageaccount).

### Copyright & Contact

©2015 by Faddah Steve Yuetsu Wolf

* [faddah on Github](https://github.com/faddah)
* [@yuetsu on Twitter](http://twitter.com/yuetsu)

*Special Thanks!*

* *thanx first and foremost to @chalkers of [Team Treehouse](http://teamtreehouse.com/) and his [very good beginner module on node.js there](http://teamtreehouse.com/library/nodejs-basics), from which i got the base code to call to APIs like [Forecast.io](http://forecast.io) and [Geonames.org](http://geonames.org) and get JSON data objects back with which you can work.*
* *the folks in the Freenode IRC #pdxnode chat channel, especially @jerrysv, @justinabrahm & @chrisdickinson, @pdxleif from @pdxnode — for how to properly parse the location JSON data object that had an array in it, chris for showing how to use the node.js 'http'* ```ready``` *keyword to return the location object properly, and @pdxleif for reminding me it's* ```process.argv.slice(2)```*, not* ```process.argv.slice[2]``` *in my code. Thanks guys!*
* *extra special thanks to my node.js mentor, @s5fs (adam), for getting my MIND RIGHT about call backs, and their use in node.js and why they don't do returns like regular JavaScript functions.*

CHANGELOG.md
## Change Log

### upcoming (2015/02/26 06:20 +00:00)
- [eb30d34](https://github.com/faddah/node-forecast-io-serve/commit/eb30d34d98becf7b961bbe4ba0f1da62117cb499) Put in appropriate line breaks in console.log() print outs. Error messages from servers forecast.io and geonames.org now correct and consistent. Updated README.md. Regenerated package.json with npm init. (@faddah)
- [dbc400c](https://github.com/faddah/node-forecast-io-serve/commit/dbc400c38c8d9a4a7cf283a756002597caa95b95) minor edit changes to README.md - slice[2] -> slice(2) where applicable. (@faddah)
- [01a7eec](https://github.com/faddah/node-forecast-io-serve/commit/01a7eecbfc658fb883f703ff74966b89cd973d38) Now all working! forecast.js calls everything and returns correct locations from zip-convert.js and zip-forecast.js, with messages in messages.js module. In messages.js, in printForecastInfo(), if returned weather object from forecast.io has a special weather alert, that is now printed out. Included in messages.js for printForecastInfo() functions to convert hPA pressure to more commone U.S. Barometric pressure inHG (inches Mercury). Converted humidity to a percentage read out. Took \'%\' off of final visability rating. Unicode characters (percentage, degrees farenheit, etc.) now work o.k. from command line, just have to check terminal character encoding set to some UTF. forecast.js now has commented out optional code for using forEach() instead of async.eachSeries() module (does not work as well, faster, but not in desired order). package.json updated using npm init. .flowconfig updated to include all *.js files. README.md edited to include all this info, thank you\'s to those who helped, and punched up the instructions on how to run it so they were more clear. (@faddah)
- [55203c2](https://github.com/faddah/node-forecast-io-serve/commit/55203c29273798c182101aca2ddf1b69f9ccf5c2) Version 0.1.0! All now working! forecast.js now is the main file that pulls all modules together and runs with node from command line, messages.js working and now has new feature of printing out special weather alerts if available, zip-convert.js & zip-forecast.js working just fine as modules feeding into forecast.js. very happy. (@faddah)
- [96e4ef4](https://github.com/faddah/node-forecast-io-serve/commit/96e4ef474cd58a00f0f3ec51d71b9e410e14e2f1) Merged master fixed conflict. (@faddah)
- [0c33b24](https://github.com/faddah/node-forecast-io-serve/commit/0c33b24856e58540e8978415461143cbaa61f369) made sure README.md matched all other branch versions & committed it. (@faddah)
- [8689bf6](https://github.com/faddah/node-forecast-io-serve/commit/8689bf645600ff5c110955101a8f5ee67dca71da) changed testing in zip-forecast.js back to just printing simple, current weather message rather than barfing out the whold damn forecast.io weather JSON object, formatting changes to correct README.md. (@faddah)
- [db17525](https://github.com/faddah/node-forecast-io-serve/commit/db17525ba7be540790aa3c16af6d3fe6ff21e8c9) Merged master fixed conflict. (@faddah)
- [00b0f0c](https://github.com/faddah/node-forecast-io-serve/commit/00b0f0cac26bc98b8967ab2654e7138fe23cce5e) Merged master fixed conflict. (@faddah)
- [2fc5c34](https://github.com/faddah/node-forecast-io-serve/commit/2fc5c34006802f8ee40cd59af1d332f6fa08eead) minor edit clean-ups to zip-convert.js & zip-forecast.js (@faddah)
- [ead662b](https://github.com/faddah/node-forecast-io-serve/commit/ead662b212af71d1fd84154153445f4ecb24e68c) added commented example data at EOF from new york of what forecast.io returns as a JSON object with all current weather data for a location. (@faddah)
- [5a85aa4](https://github.com/faddah/node-forecast-io-serve/commit/5a85aa48b3dac94ab1402c9b6eefbf5d640c68fd) cleaned up messages.js, zip-convert.js (now also exports showLocation()), and zip-forecast.js (now also exports showForecast()). (@faddah)
- [c844b06](https://github.com/faddah/node-forecast-io-serve/commit/c844b06dd4ca889b5935f8fface857ff966c10e9) added .gitignore file, mainly for *-old.js files (@faddah)
- [af53458](https://github.com/faddah/node-forecast-io-serve/commit/af53458cbfc2a2858d89de0e9df080bfc04be022) got zip-forecast.js properly working with test longitude & latitude data, created git repo on nitrous to be able to pull and push down & up to local repo via github, updated package.json using \'npm init.\' thanx to adam, @s5fs for help along the way. (@faddah)
- [1eeaa08](https://github.com/faddah/node-forecast-io-serve/commit/1eeaa0885e13489da55a235d2078a0813f63f6f8) addded .flowconfig for flow type-checking, zip-convert.js fully working, zip-forecast halfway working with module from zip-convert.js, messages.js fully working, (@faddah)
- [a8fe2fc](https://github.com/faddah/node-forecast-io-serve/commit/a8fe2fc2b8bae622efe88ea331cc8b5201fb57c7) added modular messages.js file for printing out the messages to console, zip-convert working fine, testing out why getting weird results from imported zip-convert.js in zip-forecast.js. (@faddah)
- [aa71e16](https://github.com/faddah/node-forecast-io-serve/commit/aa71e16b8ebb0a2c5a4befa7b4448f7097236629) even further formatting updates to README.md. (@faddah)
- [0b1865f](https://github.com/faddah/node-forecast-io-serve/commit/0b1865f4e187bb648ac68784e5a30273a97655ae) further formatting updates to README.md. (@faddah)
- [7a12ff8](https://github.com/faddah/node-forecast-io-serve/commit/7a12ff870579ca16fd206b07bcf22804dd9c534f) formatting updates to README.md. (@faddah)
- [78784a5](https://github.com/faddah/node-forecast-io-serve/commit/78784a561402861dfc43955a1369270776a4ec72) zip-convert.js now fully functional and returning properly the location as an JSON object, including longitude & latitude within, reduced down code in zip-forecast.js for testing, updated README.md with all this info. (@faddah)

test.js
var
  zipconvert = require('./zip-convert.js');

var zip = '97214';
var doNextThing = function (error, data) {
  console.log(error);
  console.log('Data is');
  console.log(data);
  
  // NOW we
}

var foo;
foo = zipconvert.getLocation(zip);
mydata = getWeatherData(foo);
console.log(mydata); // real weather data

zipconvert.getLocation(zip, function (error, latlong) {
  getWeatherData(latlong, function (error, mydata) {
    getMoredata(mydata, function (error, otherData) {})
    console.log(mydata);
  });
});




var worker = function (data, cb) {
  data = data + ' FOOOOO';
  
  var
    error = null,
    output = {
      data: data,
      sandwich: 'ham'
    };

  if (cb && typeof cb === 'function') {
    cb(error, output);
  }
}


worker('Adam says', function (error, output) {
  console.log(output);
});




















messages.js
/* @flow */
// Prints out the location message
function printLocationInfo(zipCode, city, state, country, long, lat) {
  var zipMessage = "Your results have been found!  Your weather forecast is below: \n";
  zipMessage += "------------------------------------------\n\n";
  zipMessage += "-- The city for the U.S. Zip Code " + zipCode + " is: " + city + ", " + state +  ", " + country + ", longitude: " + long + ", latitude: " +  lat  + ".\n";
  console.log(zipMessage);
}

// used in printForecastMessage() to assign compass degree number to general compass heading
var windDirection = function(bearing) {
  if (bearing >= 357.0 && bearing <= 3.0){
    return "N";
  } else if(bearing >= 3.01 && bearing <= 38.50) {
    return "NNW";
  } else if(bearing >= 38.51 && bearing <= 51.50) {
    return "NW"
  } else if(bearing >= 51.51 && bearing <= 87.0) {
    return "WNW";
  } else if(bearing >= 87.01  && bearing <= 93.0) {
    return "W";
  } else if(bearing >= 93.01 && bearing <= 128.50) {
    return "WSW";
  } else if(bearing >= 128.51 && bearing <= 141.50) {
    return "SW";
  } else if(bearing >= 141.51 && bearing <= 177.0) {
    return "SSW";
  } else if(bearing >= 177.01 && bearing <= 183.0) {
    return "S"
  } else if(bearing >= 183.01 && bearing <= 218.50) {
    return "SSE";
  } else if(bearing >= 218.51 && bearing <= 231.50) {
    return "SE";
  } else if(bearing >= 231.51 && bearing <= 267.0) {
    return "ESE";
  } else if(bearing >= 267.0 && bearing <= 273.0) {
    return "E";
  } else if(bearing >= 273.01 && bearing <= 308.50) {
    return "ENE";
  } else if(bearing >= 308.51 && bearing <= 321.50) {
    return "NE";
  } else {
    return "NNE";
  }
}

// used with printForecastMessage to convert hPA Atmospheric Pressure units to more common, U.S. Barometric Pressure inHG (inches Mercury).
var hPAtoinHGConversion = function(hPaUnits){
  return (hPaUnits / 33.8638866667);
}

var changeToPercent = function(float) {
  return float * 100;
}

// Prints out the final weather forecast message
function printForecastMessage(forecast, location) {
  var fCurr = forecast.currently;
  var emDash = '\u2014';
  var degF = '\u2109';
  var percent = '\uFF05';
  var forecastMsg = "-- The current weather forecast for " + location.placeName + ", " + location.adminName1 + ", " + location.countryCode + ", is: " + fCurr.summary + " " + fCurr.icon + ", temperature: " + fCurr.temperature + degF + ", feels like: " + fCurr.apparentTemperature + degF + ", wind speed: " + fCurr.windSpeed + "MPH out of the " + windDirection(fCurr.windBearing) + ", precipitation of: " + fCurr.precipIntensity + ", humidity of: " + changeToPercent(fCurr.humidity) + percent + ", a barometric pressure of: " + hPAtoinHGConversion(fCurr.pressure).toFixed(2) + " inHG (inches Mercury), and a visibility of: " + fCurr.visibility + ".\n";
  console.log(forecastMsg);
  if(forecast.alerts) {
    var fAlert = forecast.alerts[0];
    alertMsg = "There is also a Special Weather Alert Message for this region " + emDash + "\n" + fAlert.title + ": \n\n" + fAlert.description + "\nMore information can be found at this web address: " + fAlert.uri + ".\n\n";
    console.log(alertMsg);
  }
}

// Prints out error messages
function printError(error) {
  if (error) {
    console.error(error.message);
  }
}

module.exports.printLocationInfo = printLocationInfo;
module.exports.printForecastMessage = printForecastMessage;
module.exports.printError = printError;


zip-convert.js
/* @flow */
//  Problem: forecast.io only will take arguments in longitude and latitude, and input from users will be zip/postal codes.
//
//  Solution: use node going to geonames.org server api to get back JSON with longitude and latitude for given zip code
//  send user input u.s. zip code to geonames server via node

var http = require("http");
var messages = require("./messages.js");
var zip = process.argv.slice(2);

var printError = messages.printError;
var printLocationInfo = messages.printLocationInfo;

function getLocation(zip, ready) {
  //  Connect to the API URL ("http://api.geonames.org/postalCodeLookupJSON?postalcode=[ZIP]&country=US&username=[USERNAME]")
  var username = "faddah";
  var geonamesAPIURL = "http://api.geonames.org/postalCodeLookupJSON?postalcode=" + zip + "&country=US&username=" + username;
  var request = http.get(geonamesAPIURL, function(response) {
    var finalData = "";
    //  Read the data
    response.on('data', function (dataStream) {
      finalData += dataStream;
    });
    response.on('end', function() {
      // if we have a functional server connection...
      if(response.statusCode === 200 && finalData) {
        try {
    			// Parse the returned finalData object, knowing it's first key is an array
          // (corrected below, per @jerrysv & @justinabrahms of Freenode #pdxnode - thanx guys!)
          var location = JSON.parse(finalData).postalcodes[0];
          // Per @chrisdickinson (thanks, chris!), use 'ready' to return the location data object, including longitude (location.lng) and latitude (location.lat).
          ready(null, location);
        } catch(error) {
          // Parse Error
          printError(error);
        }
      } else {
        // Status Code Error
        printError({message: "There was an error getting the locaton for this zip code from the geonames.org server for \"" + zip + ",\" this may be a server error or a zip code that may not exist in the U.S. (Status Code Error: \'" + response.statusCode + " - " + http.STATUS_CODES[response.statusCode] + "\')"});
      }
    });
  });
}

var showLocation = function(error, location) {
  if(location){
    try {
      printLocationInfo(location.postalcode, location.placeName, location.adminName1, location.countryCode, location.lng, location.lat);
      return location;
    } catch(error) {
      // Locaton object null or undefined error
      printError(error);
    }
  } else {
    // Status Code Error
    printError({message: "There was an error getting the locaton for this zip code from the geonames.org server for \"" + zip + ",\" this may be a server error or a zip code that may not exist in the U.S. (Status Code Error: \'" + response.statusCode + " - " + http.STATUS_CODES[response.statusCode] + "\')"});
  }
};

// export module for use in zip-forecast.js
module.exports.getLocation = getLocation;
module.exports.showLocation = showLocation;

// example api url call to geonames for zip code:  http://api.geonames.org/postalCodeLookupJSON?postalcode=97215&country=US&username=faddah

/* * * * * * * * * * * * * * * *

Example returned JSON data from geonames.com —

{ postalcodes:
   [ { adminName2: 'Multnomah',
       adminCode2: '051',
       postalcode: '97215',
       adminCode1: 'OR',
       countryCode: 'US',
       lng: -122.599001,
       placeName: 'Portland',
       lat: 45.514282,
       adminName1: 'Oregon' } ] }

* * * * * * * * * * * * * * * */


forecast.js
/* @flow */
//
// Called like
// $> node forecast.js 97214
//  - or -
// $> node forecast.js 10011 20050 91308 90069 94101 97215

var async = require('async'),
    getLocation = require('./zip-convert').getLocation,
    showLocation = require('./zip-convert').showLocation,
    getForecast = require('./zip-forecast').getForecast,
    showForecast = require('./zip-forecast').showForecast,
    zipcodes = process.argv.slice(2),
    printError = require('./messages').printError;

//Search Initiator (Let user know we are in the process of searching)
console.log("------------------------------------------\n"+"\n" + "Searching for your weather update for "+zipcodes +"..." +"\n \n ");

// for async module call, the middle iterator argument call, calling all other modules to do THE BIG WORK.
var iterator = function (zipcode, next) {
 	// Do something w/the zipcode.
  getLocation(zipcode, function (error, location) {
    showLocation(error, location);
    getForecast(location, function (error, forecast) {
      showForecast(error, forecast, location);
      next(null, forecast);
    })
  });
 };

// for async mdule eachSeries() call, last "done" argument call.
var allDone = function (err) {
  printError(err);
  console.log('------------------------------------------\n...and those are your forecasts for today!\n');
}

 // Loop over each zipcode, one after another.
async.eachSeries(zipcodes, iterator, allDone);


/*  *  *  *  *  *  *  *  *  *

// optional other way to do it with forEach() — works, faster, but does not work as well because: not all in order
zipcodes.forEach(function (zipcode) {
  getLocation(zipcode, function (error, location) {
    showLocation(error, location);
    getForecast(location, function (error, forecast) {
      showForecast(error, forecast, location);
    });
  });
});

*  *  *  *  *  *  *  *  *  */
